### ISSUE TITLE
Index Error

### ISSUE THREAD

**Issue by MSA8D8** [goodshawn12/REST#5]
Hello Mr.Goodshawn12

I want to use your REST program made in Matlab and tried to run the test_script.m, but some index errors are happened in it. The Matlab version is R2018a, and My PC is MacBook Pro 2014 Mid.

Please give me your help. Thank you in advance.

What was happened and the error message is below. 

>> test_script
starting BCILAB...
running config script /Users/Downloads/REST-master/dependencies/BCILAB/bcilab_config.m...
running startup function...
Loading BCILAB dependencies...
Loading /Users/Downloads/REST-master/dependencies/BCILAB/dependencies/eeglab13_4_4b...
eeglab: options file is ~/eeg_options.m
EEGLAB: adding "dipfit" v2.3 (see >> help eegplugin_dipfit)
EEGLAB: adding "firfilt" v1.6.1 (see >> help eegplugin_firfilt)
Loading /Users/Downloads/REST-master/dependencies/BCILAB/dependencies/cvx-1.21.b795...
Loading /Users/Downloads/REST-master/dependencies/BCILAB/dependencies/SIFT-private...
Initializing SIFT...
/Users/Downloads/REST-master/dependencies/BCILAB/dependencies/SIFT-private
SIFT: Adding subfolders.
Start SIFTing!

>> code is in /Users/Downloads/REST-master/dependencies/BCILAB/code
data is in /Users/Downloads/REST-master/dependencies/BCILAB/userdata
results are in /Users/Downloads/REST-master/dependencies/BCILAB/userdata
cache is disabled
temp is in /Users/Downloads/REST-master/dependencies/BCILAB-temp

>> Welcome to the BCILAB toolbox on MacBook-Pro.local!
 Launch time was: 17.0 seconds

>> pop_loadset(): loading file ./data/Quick30_Shawn_EyeOpen.set ...
Reading float file './data/Quick30_Shawn_EyeOpen.fdt'...
Loading LSL library...
Creating data streaminfo...
Opening data outlet...
Preparing marker map...
Now playing back...
eeg_insertbound(): 1 boundary (break) events added.
readlocs(): 'BESA' format assumed from file extension
BESA header detected, skipping three lines...
Readlocs: BESA spherical coords. converted, now deleting BESA fields
          to avoid confusion (these fields can be exported, though)
 
Index is out of the array size.

Error: lslin (line 7)
inlet = lsl_inlet(result{1});

Error: REST>startup_initializeORICA (line 533)
lslin(handles)

Error: REST>REST_OpeningFcn (line 156)
handles = startup_initializeORICA(handles,calibration_data,config);

Error: gui_mainfcn (line 220)
    feval(gui_State.gui_OpeningFcn, gui_hFigure, [], guidata(gui_hFigure), varargin{:});

Error: REST (line 71)
    gui_mainfcn(gui_State, varargin{:});

Error: test_script (line 78)
REST(opts)

**Comment by corscheid** [comment]
Can confirm that I'm getting the exact same error.

Matlab R2018a, macOS Mojave 10.14.2 on MacBook Pro Mid 2015.

```
>> test_script
starting BCILAB...
running config script /Users/corey/Documents/GitHub/REST/dependencies/BCILAB/bcilab_config.m...
running startup function...
  The previously loaded EEGLAB path has been replaced.
Loading BCILAB dependencies...
Loading /Users/corey/Documents/GitHub/REST/dependencies/BCILAB/dependencies/eeglab13_4_4b...
eeglab: options file is ~/eeg_options.m
EEGLAB: adding "dipfit" v2.3 (see >> help eegplugin_dipfit)
EEGLAB: adding "firfilt" v1.6.1 (see >> help eegplugin_firfilt)
Loading /Users/corey/Documents/GitHub/REST/dependencies/BCILAB/dependencies/cvx-1.21.b795...
Loading /Users/corey/Documents/GitHub/REST/dependencies/BCILAB/dependencies/SIFT-private...
Initializing SIFT...
/Users/corey/Documents/GitHub/REST/dependencies/BCILAB/dependencies/SIFT-private
SIFT: Adding subfolders.
Start SIFTing!


code is in /Users/corey/Documents/GitHub/REST/dependencies/BCILAB/code
data is in /Users/corey/Documents/GitHub/REST/dependencies/BCILAB/userdata
results are in /Users/corey/Documents/GitHub/REST/dependencies/BCILAB/userdata
cache is disabled
temp is in /Users/corey/Documents/GitHub/REST/dependencies/BCILAB-temp

 Welcome to the BCILAB toolbox on css-mbp.local!
 Launch time was: 14.9 seconds

pop_loadset(): loading file ./data/Quick30_Shawn_EyeOpen.set ...
Reading float file './data/Quick30_Shawn_EyeOpen.fdt'...
Loading LSL library...
Creating data streaminfo...
Opening data outlet...
Preparing marker map...
Now playing back...
eeg_insertbound(): 1 boundary (break) events added.
readlocs(): 'BESA' format assumed from file extension
BESA header detected, skipping three lines...
Readlocs: BESA spherical coords. converted, now deleting BESA fields
          to avoid confusion (these fields can be exported, though)
Index exceeds array bounds.

Error in lslin (line 7)
inlet = lsl_inlet(result{1});

Error in REST>startup_initializeORICA (line 533)
lslin(handles)

Error in REST>REST_OpeningFcn (line 156)
handles = startup_initializeORICA(handles,calibration_data,config);

Error in gui_mainfcn (line 220)
    feval(gui_State.gui_OpeningFcn, gui_hFigure, [], guidata(gui_hFigure), varargin{:});

Error in REST (line 71)
    gui_mainfcn(gui_State, varargin{:});

Error in test_script (line 78)
REST(opts)
 
>> 
```

It looks like there is some sort of array index out of bounds error going on. Any word on what to do about this?

**Comment by goodshawn12** [comment]
Thanks for reporting the issue! The error would occur when the LSL (Lab Streaming Layer) stream was not found (and "result" variable was empty). We are guessing the source of the problem could be compatibility issue of LSL or BCILAB with MATLAB 2018a. If you have an order version of MATLAB, could you try if REST can work? In the meantime we will try to address the problem and get back to you too. 
Thanks,
Shawn

**Comment by corscheid** [comment]
I managed to solve this issue (but another one has come up, which I will post separately)

Here is what I did: 

Start by removing all MATLAB Path entries containing `REST` and `.bcilab`.

Open Terminal and either cd to a directory where REST has not been cloned, or delete REST folder before proceeding.

Then:
```
# clone REST
git clone https://github.com/goodshawn12/REST

# clone liblsl-Matlab from labstreaminglayer
git clone https://github.com/labstreaminglayer/liblsl-Matlab

# remove the liblsl-Matlab that comes with REST
rm -rf REST/dependencies/BCILAB/dependencies/liblsl-Matlab

# replace it with the newly cloned liblsl-Matlab
cp -r liblsl-Matlab REST/dependencies/BCILAB/dependencies/
```

After this, in MATLAB, the REST test_script runs as described in the README: a pipline window with checkboxes opens and once that is closed, the main REST window opens.

**Comment by goodshawn12** [comment]
Hi corscheid,
Wonderful! Thanks for testing and providing a solution. 
It then seems that the LSL library (liblsl-matlab) needs to be updated since the current one in the dependencies is not compatible with newer version of MATLAB. (This has always been an issue for LSL, BCILAB, EEGLAB, etc.)

**Comment by MSA8D8** [comment]
Hi @goodshawn12 and @corscheid 

Thanks for solving my problem. 
I have tried the solution @corscheid suggested before.
After this, in MATLAB, the REST test_script runs as described in the README, too.

But, I found another problem below.
The RSET window is never opened in my Mac.

>> test_script
starting BCILAB...
running config script /Users/.../Documents/MATLAB/REST/dependencies/BCILAB/bcilab_config.m...
running startup function...
Loading BCILAB dependencies...
Compiling the function/class lsl_append_child...success.
Compiling the function/class lsl_append_child_value...success.
Compiling the function/class lsl_append_copy...success.
Compiling the function/class lsl_child...success.
Compiling the function/class lsl_child_value...success.
Compiling the function/class lsl_child_value_n...success.
Compiling the function/class lsl_close_stream...success.
Compiling the function/class lsl_create_inlet...success.
Compiling the function/class lsl_create_outlet...success.
Compiling the function/class lsl_create_streaminfo...success.
Compiling the function/class lsl_destroy_inlet...success.
Compiling the function/class lsl_destroy_outlet...success.
Compiling the function/class lsl_destroy_streaminfo...success.
Compiling the function/class lsl_empty...success.
Compiling the function/class lsl_first_child...success.
Compiling the function/class lsl_freelib_...success.
Compiling the function/class lsl_get_channel_count...success.
Compiling the function/class lsl_get_channel_format...success.
Compiling the function/class lsl_get_created_at...success.
Compiling the function/class lsl_get_desc...success.
Compiling the function/class lsl_get_fullinfo...success.
Compiling the function/class lsl_get_hostname...success.
Compiling the function/class lsl_get_info...success.
Compiling the function/class lsl_get_name...success.
Compiling the function/class lsl_get_nominal_srate...success.
Compiling the function/class lsl_get_session_id...success.
Compiling the function/class lsl_get_source_id...success.
Compiling the function/class lsl_get_type...success.
Compiling the function/class lsl_get_uid...success.
Compiling the function/class lsl_get_version...success.
Compiling the function/class lsl_get_xml...success.
Compiling the function/class lsl_have_consumers...success.
Compiling the function/class lsl_is_text...success.
Compiling the function/class lsl_last_child...success.
Compiling the function/class lsl_library_version...success.
Compiling the function/class lsl_loadlib_...success.
Compiling the function/class lsl_local_clock...success.
Compiling the function/class lsl_name...success.
Compiling the function/class lsl_next_sibling...success.
Compiling the function/class lsl_next_sibling_n...success.
Compiling the function/class lsl_open_stream...success.
Compiling the function/class lsl_parent...success.
Compiling the function/class lsl_prepend_child...success.
Compiling the function/class lsl_prepend_child_value...success.
Compiling the function/class lsl_prepend_copy...success.
Compiling the function/class lsl_previous_sibling...success.
Compiling the function/class lsl_previous_sibling_n...success.
Compiling the function/class lsl_protocol_version...success.
Compiling the function/class lsl_pull_chunk_d...success.
Compiling the function/class lsl_pull_sample_buf...success.
Compiling the function/class lsl_pull_sample_d...success.
Compiling the function/class lsl_pull_sample_str...success.
Compiling the function/class lsl_push_chunk...success.
Compiling the function/class lsl_push_sample...success.
Compiling the function/class lsl_remove_child...success.
Compiling the function/class lsl_remove_child_n...success.
Compiling the function/class lsl_resolve_all_...success.
Compiling the function/class lsl_resolve_bypred_...success.
Compiling the function/class lsl_resolve_byprop_...success.
Compiling the function/class lsl_samples_available...success.
Compiling the function/class lsl_set_child_value...success.
Compiling the function/class lsl_set_name...success.
Compiling the function/class lsl_set_value...success.
Compiling the function/class lsl_time_correction...success.
Compiling the function/class lsl_value...success.
Compiling the function/class lsl_wait_for_consumers...success.
All files in /Users/.../Documents/MATLAB/REST/dependencies/BCILAB/dependencies/liblsl-Matlab/mex compiled successfully.
Loading /Users/.../Documents/MATLAB/REST/dependencies/BCILAB/dependencies/eeglab13_4_4b...
eeglab: options file is ~/eeg_options.m
EEGLAB: adding "dipfit" v2.3 (see >> help eegplugin_dipfit)
EEGLAB: adding "firfilt" v1.6.1 (see >> help eegplugin_firfilt)
Loading /Users/.../Documents/MATLAB/REST/dependencies/BCILAB/dependencies/cvx-1.21.b795...
Loading /Users/.../Documents/MATLAB/REST/dependencies/BCILAB/dependencies/SIFT-private...
Initializing SIFT...
/Users/.../Documents/MATLAB/REST/dependencies/BCILAB/dependencies/SIFT-private
SIFT: Adding subfolders.
Start SIFTing!
code is in /Users/.../Documents/MATLAB/REST/dependencies/BCILAB/code
data is in /Users/.../Documents/MATLAB/REST/dependencies/BCILAB/userdata
results are in /Users/.../Documents/MATLAB/REST/dependencies/BCILAB/userdata
cache is disabled
temp is in /Users/.../Documents/MATLAB/REST/dependencies/BCILAB-temp

 Welcome to the BCILAB toolbox on MacBook-puro.local!
 Launch time was: 57.5 seconds


Error using lsl_loadlib (line 62)
**Apparently the file "/Users/.../Documents/MATLAB/REST/dependencies/BCILAB/dependencies/liblsl-Matlab/bin/liblsl64.dylib" is missing on your computer**. Cannot load the lab streaming layer.

Error in REST>REST_OpeningFcn (line 115)
handles.lsllib = lsl_loadlib(env_translatepath('dependencies:/liblsl-Matlab/bin'));

Error in gui_mainfcn (line 220)
    feval(gui_State.gui_OpeningFcn, gui_hFigure, [], guidata(gui_hFigure), varargin{:});

Error in REST (line 71)
    gui_mainfcn(gui_State, varargin{:});

Error in test_script (line 100)
REST(opts)
 

I wonder why the bold letter in the Error message can happen. 
I tried to see what kind of files are in the path, but nothing but only readme.txt was there.
The readme.txt says "Please make sure that you have the most recent binaries from liblsl/bin/ in here."

Could you please give me any good ideas to solve this problem?
Thank you in advance.

**Comment by MSA8D8** [comment]
I can solve the problem by downloading the libls64.dylib from the link below.

https://github.com/chkothe/pylsl/blob/master/pylsl/liblsl64.dylib

Here is the message from MATLAB after running the test_script.m
And, I finally see the REST window opens and works well.

I hope that HD-72 made by Cognionics will be available to use RSET script @goodshawn12 has made.

Thank you for everything, @goodshawn12 and @corscheid.

>> test_script
pop_loadset(): loading file ./data/Smarting_220219_1311.set ...
Reading float file './data/Smarting_220219_1311.fdt'...
Loading LSL library...
Creating data streaminfo...
Opening data outlet...
Preparing marker map...
Now playing back...
Upper time limits exceed data, corrected
readlocs(): 'BESA' format assumed from file extension
BESA header detected, skipping three lines...
Readlocs: BESA spherical coords. converted, now deleting BESA fields
          to avoid confusion (these fields can be exported, though)
WARNING: The value assigned to srate in set_new is not in the typical range [0.1 100000], but was: 0
  using montage resources:/caps/Standard-10-5-Cap385.cap.
Using deprecated argument "flag_quantile/FlaggedQuantile" in function flt_clean_windows (help: Legacy parameter for pre-2012 methods. Quantile of data windows flagged for removal. Windows are emoved if flagged in all except for some possibly bad channels, controls the aggressiveness of the rejection.).
WARNING: The MVAR algorithm 'BSBL BO' depends on BSBL_BO.m, which cannot be located on the path. This algorithm will not be available.
WARNING: The MVAR algorithm 'BSBL L1' depends on BSBL_L1_noise.m, which cannot be located on the path. This algorithm will not be available.
WARNING: The MVAR algorithm 'T-MSBL' depends on TMSBL.m, which cannot be located on the path. This algorithm will not be available.
WARNING: The MVAR algorithm 'SCSA EM' depends on scsa_em.m, which cannot be located on the path. This algorithm will not be available.
Note: The head model named "resources:/headmodels/standard-Colin27-385ch.mat" was neither found in the base workspace nor could it be loaded as a file.
WARNING: The value assigned to stddev_cutoff in flt_repair_bursts is not in the typical range [2 15], but was: 20
Using deprecated argument "num_bins/NumBins" in function @(f,a,frame__f3)feval(f,a{:}) (help: This parameter is now auto-determined).
Using deprecated argument "fit_quantiles/FitQuantiles" in function @(f,a,frame__f3)feval(f,a{:}) (help: Old Truncated Gaussian quantile. This was allowed to be a matrix holding an ensemble of alternative quantiles -- that is no longer needed/supported (instead the replacement parameter TruncateQuantile should be used).).
Using deprecated argument "min_badchans/MinAffectedChannels" in function flt_clean_windows (help: Legacy parameter for 2011 method. This is the minimum number of channels that need to be affected for the window to be considered "bad".).
Using deprecated argument "ignored_chans/MaxIgnoredChannels" in function flt_clean_windows (help: Legacy parameter for 2010 method. Maximum number or ratio of channels to ignore. These can contain arbitrary data without affecting the outcome (e.g., can be unplugged channels).).
Using deprecated argument "ref_wndlen/ReferenceWindowLength" in function flt_repair_bursts (help: Selection granularity for reference data. When clean data is extracted from the recording for calibrating the cutoff thresholds the data is chopped up into windows of this length. This parameters has been deprecated in favor of ReferenceExtraction.WindowLength.).
Using deprecated argument "ref_maxbadchannels/ReferenceMaxBadChannels" in function flt_repair_bursts (help: Maximum bad-channel fraction on reference data. The maximum fraction of bad channels per time window of the data that is used as clean reference EEG on which statistics are based. Instead of a number one may also directly pass in a data set that contains clean reference data (for example a minute of resting EEG). A lower values makes this criterion more aggressive. Reasonable range: 0.05 (very aggressive) to 0.3 (quite lax). This parameters has been deprecated in favor of ReferenceExtraction.MaxBadChannels.).
Using deprecated argument "ref_tolerances/ReferenceTolerances" in function flt_repair_bursts (help: Power tolerances on reference data. These are the power tolerances beyond which a channel in the clean reference data is considered "bad", in standard deviations relative to a robust EEG power distribution (lower and upper bound). This parameters has been deprecated in favor of ReferenceExtraction.PowerTolerances.).
Flushing data set to disk...
io_loadset(): loading /Users/masaru/Documents/MATLAB/REST/dependencies/BCILAB-temp/flushedsets/1593345551.sto...
Scanning potential marker channel 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 
Using deprecated argument "num_bins/NumBins" in function @(f,a,frame__f2)feval(f,a{:}) (help: This parameter is now auto-determined).
Using deprecated argument "fit_quantiles/FitQuantiles" in function @(f,a,frame__f2)feval(f,a{:}) (help: Old Truncated Gaussian quantile. This was allowed to be a matrix holding an ensemble of alternative quantiles -- that is no longer needed/supported (instead the replacement parameter TruncateQuantile should be used).).
Note: cannot determine free disk space on your platform, trying to cache results anyway. This message will only be shown once.
Finding a clean section of the data...
Using deprecated argument "num_bins/NumBins" in function @(f,a,frame__f1)feval(f,a{:}) (help: This parameter is now auto-determined).
Using deprecated argument "fit_quantiles/FitQuantiles" in function @(f,a,frame__f1)feval(f,a{:}) (help: Old Truncated Gaussian quantile. This was allowed to be a matrix holding an ensemble of alternative quantiles -- that is no longer needed/supported (instead the replacement parameter TruncateQuantile should be used).).
Determining time window rejection thresholds...done.
Keeping 79.3% (56 seconds) of the data.
Estimating statistics...
Determining per-component thresholds...done.

**Comment by lucapton** [comment]
Was libls64.dylib no present in that folder before you downloaded it before? It should have been as it's included in the github dependices: https://github.com/goodshawn12/REST/tree/master/dependencies/BCILAB/dependencies/liblsl-Matlab/bin

**Comment by MSA8D8** [comment]
@lucapton 
The folder I used for liblsl64.dylib was from the link @corscheid suggested.
> # clone liblsl-Matlab from labstreaminglayer
git clone https://github.com/labstreaminglayer/liblsl-Matlab

**Comment by corscheid** [comment]
Ahh that's right! I forgot I also did something else. See the readme in the
bin folder of labstreaminglayer/liblsl-Matlab repository, where the dylib
should be: [
https://github.com/labstreaminglayer/liblsl-Matlab/blob/master/README.md](https://github.com/labstreaminglayer/liblsl-Matlab/blob/master/README.md)

What I believe I did was compile everything from source in
labstreaminglayer repo. Results in the needed dylib. I forgot that the
binary is not contained in the git repo (or it's Matlab submodule) along
with the source code. I'll post the exact commands I used to compile and
copy when I'm able. Sorry about that!
